# 00. Table of Contents

01. Import Libraries
02. Import Data

# 01. Import Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Import Data

In [2]:
# Import products.csv as df_prods
path = r'/Users/bertpearce/Documents/Python Projects/2024.05.22 Instacard Basket Analysis/'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# Import orders_wrangled.csv as df_ords
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

## Step 1

"If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now."

In [3]:
# Review df_ords stats with the df.describe() function
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

## Step 2
"Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further."

The max order number is "100" even though there are 3.4 million records.

In [4]:
# Create a dataframe

df_test = pd.DataFrame()

In [5]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [6]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Check for mixed types in df_test

for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [11]:
# Fix df_test mixed types; Convert column mix to string

df_test['mix'] = df_test['mix'].astype('str')

In [14]:
# Confirm df_test column 'mix' is type 0 aka string

df_test['mix'].dtype

dtype('O')

## Step 3

"Check for mixed-type data in your df_ords dataframe."

In [17]:
# Review df_ords column types using dtypes

df_ords[['order_id', 'user_id', 'order_number', 'orders_day_of_week', 'order_hour_of_day', 'days_since_prior_order']].dtypes

order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [24]:
# Validate df_ords column types; Print mixed type columns

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:    
    print (col)
  else:
    print(f'No mixed data in column {col}.')

No mixed data in column Unnamed: 0.
No mixed data in column order_id.
No mixed data in column user_id.
No mixed data in column order_number.
No mixed data in column orders_day_of_week.
No mixed data in column order_hour_of_day.
No mixed data in column days_since_prior_order.


## Step 4

"If you find mixed-type data, fix it. The column in question should contain observations of a single data type."

No mixed-type data in columns.

In [25]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [26]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [28]:
# Print and review records with null values from one column

df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [29]:
# Review df_prods dimensions

df_prods.shape

(49693, 5)

In [30]:
# Create df_prods_clean by filtering out records with null values in column product_name

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [31]:
# Review df_prods dimensions

df_prods_clean.shape

(49677, 5)

## Step 5.1

Run a check for missing values in your df_ords dataframe.

In [32]:
# Identify columns with null observations; Print sums of null observations by column

df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [33]:
# Identify null observations; Print records with null values from one column

df_ords_days_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

df_ords_days_nan.head(20)

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
11           11   2168274        2             1                   2   
26           26   1374495        3             1                   1   
39           39   3343014        4             1                   6   
45           45   2717275        5             1                   3   
50           50   2086598        6             1                   5   
54           54   2565571        7             1                   3   
75           75    600894        8             1                   6   
79           79    280530        9             1                   1   
83           83   1224907       10             1                   2   
89           89   2411567       11             1                   0   
97           97   1483446       12             1                   1   
103         103   2618231       13             1                   6   
116         116   2906263       14             1                   2   
130         130   2514324       15             1                   5   
153         153   1273676       16             1                   2   
160         160   1737705       17             1                   2   
201         201   2780889       18             1                   6   
208         208   1321974       19             1                   6   
218         218    947722       20             1                   1   

     order_hour_of_day  days_since_prior_order  
0                    8                     NaN  
11                  11                     NaN  
26                  14                     NaN  
39                  11                     NaN  
45                  12                     NaN  
50                  18                     NaN  
54                   9                     NaN  
75                   0                     NaN  
79                  17                     NaN  
83                  14                     NaN  
89                  11                     NaN  
97                  12                     NaN  
103                 12                     NaN  
116                 18                     NaN  
130                 13                     NaN  
153                  8                     NaN  
160                 13                     NaN  
201                 18                     NaN  
208                 13                     NaN  
218                 10                     NaN

In [39]:
# Count the records in df_ords where order_number = 1 and days_since_prior_order = Nan

count = df_ords[(df_ords['days_since_prior_order'].isnull()) & (df_ords['order_number'] == 1)].shape[0]

print(count)

206209


## Step 5.2
"In a markdown cell, report your findings and propose an explanation for any missing values you find."

The column days_since_prior_order has 206209 null observations. The count of records where order_number = 1 and days_since_prior_order = Nan is also 206209. Therefore, days_since_prior_order is a null value because there wasn't a previous purchase.

In [42]:
# Count the records in df_ords where days_since_prior_order = 0

count = df_ords[(df_ords['days_since_prior_order'] == 0)].shape[0]

print(count)

67755


## Step 6.1

"Address the missing values using an appropriate method."

No change needed.

## Step 6.2
"In a markdown cell, explain why you used your method of choice."

All records with null values are first purchases by customers and therefore vital to the dataset so they shouldn't be removed. As there are already observations under days_since_prior_order that equal 0, converting the NaN values to 0 or any imputed values would skew the data. A new column could be made that clearly identifies each first purchase, however that can already be identified by records where order_number equals 1. No change to df_ords is needed, but it would be useful to identify the significance of NaN in days_since_prior_order for the reader's information.

In [44]:
# Identify duplicate records; Create df_dups from duplicate records in df_prods_clean

df_dups = df_prods_clean[df_prods_clean.duplicated()]

df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [45]:
# Review df_prods_clean row count

df_prods_clean.shape

(49677, 5)

In [46]:
# Create new dataframe df_prods_clean_no_dups from df_prods_clean without the duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [48]:
# Review df_prods_clean_no_dups row count

df_prods_clean_no_dups.shape

(49672, 5)

## Step 7.1

"Run a check for duplicate values in your df_ords data."

In [50]:
# Identify duplicate records; Create df_ords_dupes of from duplicate records in DF

df_ords_dupes = df_ords[df_ords.duplicated()]

df_ords_dupes

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [53]:
# Identify duplicate values in one column that should only have unique values; Create df_ords_order_id_dupes of from duplicate records in DF and then print the new dataframe

df_ords_order_id_dupes = df_ords[df_ords.duplicated(subset='order_id')]

df_ords_order_id_dupes

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

## Step 7.2

"In a markdown cell, report your findings and propose an explanation for any duplicate values you find."

No duplicates were found in df_ords.

## Step 8.1

"Address the duplicates using an appropriate method."

No duplicates were found in df_ords.

## Step 8.2

"In a markdown cell, explain why you used your method of choice."

No duplicates were found in df_ords.

In [55]:
# Export DF to a new CSV file

# Define the folder path and file name

folder_path = r'/Users/bertpearce/Documents/Python Projects/2024.05.22 Instacard Basket Analysis/'
file_name = 'products_prepared.csv'

# Validate folder

os.makedirs(folder_path, exist_ok=True)

# Create the full file path

file_path = os.path.join(folder_path, '02 Data', 'Prepared Data', file_name)

# Export the dataframe to a CSV file

df_prods_clean_no_dups.to_csv(file_path, index=False)

print(f"DataFrame exported to {file_path}")

DataFrame exported to /Users/bertpearce/Documents/Python Projects/2024.05.22 Instacard Basket Analysis/02 Data/Prepared Data/products_prepared.csv


In [56]:
# Export DF to a new CSV file

# Define the folder path and file name

folder_path = r'/Users/bertpearce/Documents/Python Projects/2024.05.22 Instacard Basket Analysis/'
file_name = 'orders_prepared.csv'

# Validate folder

os.makedirs(folder_path, exist_ok=True)

# Create the full file path

file_path = os.path.join(folder_path, '02 Data', 'Prepared Data', file_name)

# Export the dataframe to a CSV file

df_ords.to_csv(file_path, index=False)

print(f"DataFrame exported to {file_path}")

DataFrame exported to /Users/bertpearce/Documents/Python Projects/2024.05.22 Instacard Basket Analysis/02 Data/Prepared Data/orders_prepared.csv
